ada_feeding/ada_feeding/behaviors/acquisition/compute_food_frame.py

```diff
@@ -5,56 +5,246 @@
+                                                        for debugging purposes.
+        """
+        # pylint: disable=unused-argument
+        # Arguments are handled generically in base class.
+        super().blackboard_outputs(
+            **{key: value for key, value in locals().items() if key != "self"}
+        )
+
+    # pylint: disable=attribute-defined-outside-init
+    # It is okay for attributes in behaviors to be
+    # defined in the setup / initialise functions.
+
+    def setup(self, **kwargs):
+        """
+        Middleware (i.e. TF) setup
+        """
+
+        # Create TF Tree
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(
+            self.tf_buffer, self.blackboard_get("ros2_node")
+        )
 
-    def setup(self, **kwargs) -> None:
-        pass
+    def initialise(self):
+        """
+        Behavior initialization
+        """
+
+        # Construct camera intrinsics
+        # See: https://medium.com/@yasuhirachiba/converting-2d-image-coordinates-
+        #      to-3d-coordinates-using-ros-intel-realsense-d435-kinect-88621e8e733a
+        camera_info = self.blackboard_get("camera_info")
+        self.intrinsics = pyrealsense2.intrinsics()
+        self.intrinsics.width = camera_info.width
+        self.intrinsics.height = camera_info.height
+        self.intrinsics.ppx = camera_info.k[2]
+        self.intrinsics.ppy = camera_info.k[5]
+        self.intrinsics.fx = camera_info.k[0]
+        self.intrinsics.fy = camera_info.k[4]
+        if camera_info.distortion_model == "plumb_bob":
+            self.intrinsics.model = pyrealsense2.distortion.brown_conrady
+        elif camera_info.distortion_model == "equidistant":
+            self.intrinsics.model = pyrealsense2.distortion.kannala_brandt4
+        else:
+            self.intrinsics.model = pyrealsense2.distortion.none
+        self.intrinsics.coeffs = list(camera_info.d)
 
     def update(self) -> py_trees.common.Status:
-        pass
+        """
+        Behavior tick (DO NOT BLOCK)
+        """
+        # pylint: disable=too-many-locals
+        # I think this is reasonable to understand
+        # the logic of this function.
+
+        # pylint: disable=too-many-statements
+        # We can't get around all the conversions
+        # to ROS2 msg types, which take 3-4 statements each.
+
+        camera_frame = self.blackboard_get("camera_info").header.frame_id
+        node = self.blackboard_get("ros2_node")
+
+        # Check if we have the camera transform
+        if not self.tf_buffer.can_transform(
+            self.blackboard_get("world_frame"),
+            camera_frame,
+            rclpy.time.Time(),
+        ):
+            # Not yet, wait for it
+            # Use a Timeout decorator to determine failure.
+            return py_trees.common.Status.RUNNING
+        transform = self.tf_buffer.lookup_transform(
+            self.blackboard_get("world_frame"),
```

Comment:
Shouldn't we get the world frame once before the if statement, as opposed to twice? It doesn't matter much if the world frame is constant, but if this behavior is used with the world frame being written to the blackboard, we would be doing two blackboard lookups instead of one, which seems unnecessary.

Reply:
I will say a blackboard lookup is just a hash table, so doesn't take any longer than just looking up the remapped constant value.

That said, it is probably safer in parallel trees to just do the single lookup.